packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,163 @@
+// eslint-disable-next-line import/no-internal-modules
+import semverGte from "semver/functions/gte.js";
+// eslint-disable-next-line import/no-internal-modules
+import semverLte from "semver/functions/lte.js";
+// eslint-disable-next-line import/no-internal-modules
+import semverValid from "semver/functions/valid.js";
+
+import {
+	disabledCompressionConfig,
+	enabledCompressionConfig,
+} from "./compressionDefinitions.js";
+import { type IContainerRuntimeOptionsInternal } from "./containerRuntime.js";
+import { pkgVersion } from "./packageVersion.js";
+
+/**
+ * Our policy is to support N/N-1 compatibility by default, where N is the most recent public major release of the runtime.
+ * Therefore, if the customer does not provide a compatibility mode, we will default to use N-1.
+ *
+ * However, this is not consistent with today's behavior. Some options (i.e. batching, compression) are enabled by default
+ * despite not being compatible with 1.x clients. Since the policy was introduced during 2.x's lifespan, N/N-1 compatibility
+ * by **default** will be in effect starting with 3.0. Importantly though, N/N-1 compatibility is still guaranteed with the
+ * proper configurations set.
+ *
+ */
+export const defaultCompatibilityMode = "2.0.0" as const;
+
+/**
+ * String in a valid semver format.
+ */
+export type SemanticVersion =
+	| `${number}.${number}.${number}`
+	| `${number}.${number}.${number}-${string}`;
+
+/**
+ * Generic type for docSchemaAffectingOptionsConfigMap
+ */
+export type ConfigMap<T extends Record<string, unknown>> = {
+	[K in keyof T]: {
+		[version: SemanticVersion]: T[K];
+	};
+};
+
+/**
+ * Subset of the IContainerRuntimeOptionsInternal properties which affect the DocumentSchema.
+ *
+ * @remarks
+ * When a new option is added to IContainerRuntimeOptionsInternal, we must consider if it changes the DocumentSchema.
+ * If so, then a corresponding entry must be added to `docSchemaAffectingOptionsConfigMap` below. If not, then
+ * it must be omitted from this type.
+ *
+ * Note: We use `Omit` instead of `Pick` to ensure that all new options are included in this type by default. If any new properties
+ * are added to IContainerRuntimeOptionsInternal, they will be included in this type unless explicitly omitted. This will prevent
+ * us from forgetting to account for any new properties in the future.
+ */
+export type DocSchemaAffectingRuntimeOptions = Required<
+	Omit<
+		IContainerRuntimeOptionsInternal,
+		| "chunkSizeInBytes"
+		| "maxBatchSizeInBytes"
+		| "loadSequenceNumberVerification"
+		| "summaryOptions"
+	>
+>;
+
+/**
+ * Mapping of DocSchemaAffectingRuntimeOptions to their compatibility related configs.
+ *
+ * Each key in this map corresponds to a property in DocSchemaAffectingRuntimeOptions.
+ * The value is an object that maps a version string to the default value for that property when using compatibilityMode of at least that version.
```

Comment:
Consider adding a little more description about what the key-value pairs signify. For example, if a congfig is turned on in version X, it means that all clients at version X understand the format changes introduced by that feature. And then, in the example below, you could tie it in by saying something like - clients running 2.0.0 have the code and understand the format changes added by the grouped batching feature.